app/src/pages/api/auth/[...nextauth].ts

```diff
@@ -1,4 +1,190 @@
-import NextAuth from "next-auth";
-import { authOptions } from "~/server/auth";
+import CredentialsProvider from "next-auth/providers/credentials";
+import { PrismaAdapter } from "@next-auth/prisma-adapter";
+import GitHubModule, { type GithubProfile } from "next-auth/providers/github";
+import NextAuth, { type AuthOptions, type Profile } from "next-auth";
+import type { NextApiRequest, NextApiResponse } from "next";
+import { decode, encode } from "next-auth/jwt";
+import { prisma } from "~/server/db";
+import { ensureDefaultExport } from "~/utils/utils";
+import { env } from "~/env.mjs";
+import { getCookie, setCookie } from "cookies-next";
+import { randomUUID } from "crypto";
+import { type Provider } from "next-auth/providers";
+import * as Sentry from "@sentry/nextjs";
+import { captureSignup } from "~/utils/analytics/serverAnalytics";
 
-export default NextAuth(authOptions);
+const GitHubProvider = ensureDefaultExport(GitHubModule);
+
+export function authOptionsWrapper(req: NextApiRequest, res: NextApiResponse) {
+  const isCredentialsCallback =
+    req.query?.nextauth?.includes("callback") &&
+    req.query.nextauth?.includes("credentials") &&
+    req?.method === "POST";
+  const providers: Provider[] = [
```

Comment:
Let's instead construct the providers positively. Eg

```ts
const providers: Provider[] = (env.GITHUB_CLIENT_ID && env.GITHUB_CLIENT_SECRET) ? [ /* github provider */ ] : [ /* local provider */ ]
```

Reply:
Good one. Definitely looks a lot cleaner now!